utils/chunked_vector.hh

```diff
@@ -431,18 +424,38 @@ chunked_vector<T, max_contiguous_allocation>::make_room(size_t n, bool stop_afte
 }
 
 template <typename T, size_t max_contiguous_allocation>
-void
-chunked_vector<T, max_contiguous_allocation>::do_reserve_for_push_back() {
-    if (_capacity == 0) {
-        // allocate a bit of room in case utilization will be low
-        reserve(boost::algorithm::clamp(512 / sizeof(T), 1, max_chunk_capacity()));
-    } else if (_capacity < max_chunk_capacity() / 2) {
-        // exponential increase when only one chunk to reduce copying
-        reserve(_capacity * 2);
+template <typename... Args>
+T&
+chunked_vector<T, max_contiguous_allocation>::reserve_for_emplace_back(Args&&... args) {
+    if (_capacity > _size) {
+        new (addr(_size)) T(std::forward<Args>(args)...);
     } else {
-        // add a chunk at a time later, since no copying will take place
-        reserve((_capacity / max_chunk_capacity() + 1) * max_chunk_capacity());
+        size_t last_chunk_reserve_size;
+        if (_capacity == 0) {
+            // allocate a bit of room in case utilization will be low
+            last_chunk_reserve_size = boost::algorithm::clamp(512 / sizeof(T), 1, max_chunk_capacity());
+        } else if (_capacity < max_chunk_capacity() / 2) {
+            // exponential increase when only one chunk to reduce copying
+            last_chunk_reserve_size = _capacity * 2;
+        } else {
+            // add a chunk at a time later, since no copying will take place
+            last_chunk_reserve_size = max_chunk_capacity();
+        }
+        auto dest_chunk = new_chunk(last_chunk_reserve_size);
+        auto last_chunk_size = _size % max_chunk_capacity();
+        // emplace_back before migrating existing elements
+        // since `args` might refer to an element existing
+        // in the last chunk
+        new (dest_chunk.get() + last_chunk_size) T(std::forward<Args>(args)...);
+        if (_size) {
+            migrate(_chunks.back().get(), _chunks.back().get() + last_chunk_size, dest_chunk.get());
```

Comment:
What if, before the call, last_chunk_size == 0? In this case we need to take the other branch and push_back the new chunk (which has just a single new element).

Reply:
I don't think that's possible, but we can have logic for it to be on the safe side.
In that case _capacity would be equal to _size because we fill chunks to max_chunk_size, and we would allocate another chunk only when reserving _capacity > _size.

Reply:
Algorithm changed in v2. It's more straight forward now and an existing chunk is replaced with the reallocated one, after migrating the existing elements, only when needed, by checking the existing last chunk size (_size % max_chunk_capacity).